provider_client.go

```diff
@@ -210,7 +219,15 @@ func (client *ProviderClient) Request(method, url string, options *RequestOpts)
 			}
 		case http.StatusUnauthorized:
 			if client.ReauthFunc != nil {
-				err = client.ReauthFunc()
+				if client.mut != nil {
+					client.mut.Lock()
+					if curtok := client.TokenID; curtok == prereqtok {
```

Comment:
the token is one public field and can be touched outside. It would be better to make it private and set up interface for touching token to avoid pollutions.
From this part, the token can be dirty, too, and the reauth may not work.

Reply:
The token is public for historical reasons (namely, Gophercloud predates `internal` directory magic), but applications need not (and should not) set it. If an application is setting the token on its own, it's the onus of the application to do it safely.

Reply:
However, your point is valid and is something to consider in the future if we decide to make that backward-incompatible change.

Reply:
@armstrongli A compromise could be adding public get and set functions for the token, in which contain the `ProviderClient` locking/unlocking. Would that be helpful? Or maybe that's what you're suggesting (with the addition of unexporting the token)?

Reply:
yes. exactly

Reply:
Good idea. I've added them.